module/zfs/zio.c

```diff
@@ -4424,6 +4492,100 @@ zio_alloc_zil(spa_t *spa, objset_t *os, uint64_t txg, blkptr_t *new_bp,
 	return (error);
 }
 
+/*
+ * ==========================================================================
+ * Read and write to the logical top device
+ * ==========================================================================
+ */
+/*
+ * Logical reads and writes arrive at zio_logical_io_start(), which issues the
+ * first child IO to specific vdevs (through zio_vdev_child_io()). The logical
+ * IO will wait at zio_logical_io_done() for the tree of child IO to complete,
+ * and then process its result.
+ *
+ * Thus, these represent the handoff between the logical IO pipeline
+ * (compression, encryption, checksums, DVA allocation, etc) and the physical
+ * IO pipeline (the vdev tree).
+ */
+
+static void
+zio_logical_io_child_done(zio_t *zio)
+{
+	zio_t *pio = zio->io_private;
+	pio->io_error = zio->io_error;
+}
+
+static zio_t *
+zio_logical_io_start(zio_t *zio)
+{
+	spa_t *spa = zio->io_spa;
+
+	ASSERT3U(zio->io_child_type, >, ZIO_CHILD_VDEV);
+	ASSERT3P(zio->io_bp, !=, NULL);
+	ASSERT3P(zio->io_vd, ==, NULL);
+	ASSERT3U(zio->io_error, ==, 0);
+	ASSERT3U(zio->io_child_error[ZIO_CHILD_VDEV], ==, 0);
+	ASSERT(zio->io_type == ZIO_TYPE_READ || zio->io_type == ZIO_TYPE_WRITE);
+
+	if (!(zio->io_flags & ZIO_FLAG_CONFIG_WRITER))
+		spa_config_enter(spa, SCL_ZIO, zio, RW_READER);
+
+	zio_nowait(zio_vdev_child_io(zio, zio->io_bp, spa->spa_root_vdev,
+	    zio->io_offset, zio->io_abd, zio->io_size, zio->io_type,
+	    zio->io_priority, 0, zio_logical_io_child_done, zio));
```

Comment:
Considering that `zio_t` is more than 1KB in size, aside of time to allocate and handle it, the additional ZIO layer will also increase memory pressure.  The most obvious it would be on small I/Os, like BRT/DDT, etc, so your benchmark should be better run on some very small record sizes, otherwise they may be not informative.

Reply:
Not dramatic, but what would you say about https://github.com/openzfs/zfs/pull/17619 ?